LIPS/LIP-draft_on_coverage_app.md

```diff
@@ -0,0 +1,465 @@
+    );
+```
+Emitted when the stEth `amount` (corresponding to `sharesAmount` shares) burnt for the `isCover` reason.
+
+See: `processLidoOracleReport`.
+
+### Event: ExcessStETHRecovered
+```solidity
+    event ExcessStETHRecovered(
+        address indexed requestedBy,
+        uint256 amount,
+        uint256 sharesAmount
+    );
+```
+Emitted when the excessive stEth `amount` (corresponding to `sharesAmount`) recovered (e.g. transferred) to the Lido treasure address by `requestedBy` sender.
+
+See: `recoverExcessStETH`.
+
+### Event: ERC20Recovered
+```solidity
+    event ERC20Recovered(
+        address indexed requestedBy,
+        address indexed token,
+        uint256 amount
+    );
+```
+Emitted when the ERC20 `token` recovered (e.g. transferred) to the Lido treasure address by `requestedBy` sender.
+
+See: `recoverERC20`.
+
+## Security Considerations
+
+#### The proposed contract is non-upgradable and non-ownable
+
+There are no intermediate proxies. All of the external addresses are set only on the deployment stage. In case of emergency or crucial update, we would deploy a new contract and reset `IBeaconReportReceiver` callback address on `LidoOracle`. A newly deployed replacement contract should be initialized with previously accumulated cover/non-cover counter values.
+
+#### Burning is only allowed on the `LidoStEthBurner` contract address
+
+The `BURN_ROLE` that's assigned to the `LidoStEthBurner` is restricted by the Aragon ACL parameters. That way, `Lido.burnShares()` method could only be executed for the `LidoSharesBurner` contract address by the contract itself and there are no other shares burning interfaces exists.
+
+#### There is no way to call `processOracleReport` from any address except the `LidoOracle` contract
+
+An explicit pre-condition (`require(msg.sender == Lido.getOracle()`) is checked for the message sender address.
+
+## Failure modes
+
+#### DAO decides to grant a shares burning permission to someone else
+
+It can break the proposal concept of maintaining global (protocol-wide) shares burnt counters. 
+So the `StEthSharesBurner` contract will become almost useless by providing incomplete and even incorrect information about ever burnt shares.
+
+#### Someone apply slashing coverage using non-cover request type
+
+The Anchor Protocol will receive incorrect rewards.
+
+#### Someone apply non-cover using cover request type
+
+Anchor/bETH token holders will lose some rewards. 
+
+## Reference implementation
+
+Reference implementation sketch of the `LidoStSharesBurner` interface is given below
+```solidity=
+interface IERC20;
+interface IStEth;
+interface ILido;
+
+interface IBeaconReportReceiver /* already exists in lido sources */ {
+    function processLidoOracleReport(uint256 _postTotalPooledEther,
+                                     uint256 _preTotalPooledEther,
+                                     uint256 _timeElapsed) external;
+}
+
+contract LidoStETHBurner is IBeaconReportReceiver {
```

Comment:
Good one! I have several gas golfing optimizations in mind, and several code style considerations. Let's discuss on the coding stage

Reply:
Ok, thx!